src/lsm/set_associative_cache.zig

```diff
@@ -0,0 +1,726 @@
+    p.set(2, 0b11);
+    try testing.expectEqual(@as(u64, 0b00111001), words[0]);
+    try testing.expectEqual(@as(u2, 0b11), p.get(2));
+    p.set(3, 0b11);
+    try testing.expectEqual(@as(u64, 0b11111001), words[0]);
+    try testing.expectEqual(@as(u2, 0b11), p.get(3));
+    p.set(3, 0b01);
+    try testing.expectEqual(@as(u64, 0b01111001), words[0]);
+    try testing.expectEqual(@as(u2, 0b01), p.get(3));
+    p.set(3, 0b00);
+    try testing.expectEqual(@as(u64, 0b00111001), words[0]);
+    try testing.expectEqual(@as(u2, 0b00), p.get(3));
+
+    p.set(4, 0b11);
+    try testing.expectEqual(@as(
+        u64,
+        0b0000000000000000000000000000000000000000000000000000001100111001,
+    ), words[0]);
+    p.set(31, 0b11);
+    try testing.expectEqual(@as(
+        u64,
+        0b1100000000000000000000000000000000000000000000000000001100111001,
+    ), words[0]);
+}
+
+test "PackedUnsignedIntegerArray: fuzz" {
+    const seed = 42;
+
+    var prng = std.rand.DefaultPrng.init(seed);
+    const random = prng.random();
+
+    inline for (.{ u1, u2, u4 }) |UInt| {
+        const Context = ArrayTestContext(UInt);
+
+        var context = try Context.init(random, 1024);
+        defer context.deinit();
+
+        try context.run();
+    }
+}
+
+fn BitIterator(comptime Bits: type) type {
+    return struct {
+        const Self = @This();
+        const BitIndex = math.Log2Int(Bits);
+
+        bits: Bits,
+
+        /// Iterates over the bits, consuming them.
+        /// Returns the bit index of each set bit until there are no more set bits, then null.
+        inline fn next(it: *Self) ?BitIndex {
+            if (it.bits == 0) return null;
+            // This @intCast() is safe since we never pass 0 to @ctz().
+            const index = @intCast(BitIndex, @ctz(Bits, it.bits));
+            // Zero the lowest set bit.
+            it.bits &= it.bits - 1;
+            return index;
+        }
+    };
+}
+
+test "BitIterator" {
+    const expectEqual = @import("std").testing.expectEqual;
+
+    var it = BitIterator(u16){ .bits = 0b1000_0000_0100_0101 };
+
+    for ([_]u4{ 0, 2, 6, 15 }) |e| {
+        try expectEqual(@as(?u4, e), it.next());
+    }
+    try expectEqual(it.next(), null);
+}
+
+test "SetAssociativeCache: search_tags()" {
+    const testing = std.testing;
+
+    const log = false;
+    const seed = 42;
+
+    const Key = u64;
+    const Value = u64;
+
+    const context = struct {
+        inline fn key_from_value(value: Value) Key {
+            return value;
+        }
+        inline fn hash(key: Key) u64 {
+            return key;
+        }
+        inline fn equal(a: Key, b: Key) bool {
+            return a == b;
+        }
+    };
+
+    const layout: Layout = .{};
+    const SAC = SetAssociativeCache(
+        Key,
+        Value,
+        context.key_from_value,
+        context.hash,
+        context.equal,
+        layout,
+    );
+    if (log) SAC.inspect();
+
+    const Tag = meta.Int(.unsigned, layout.tag_bits);
```

Comment:
We should test `search_tags()` exhaustively for different sized tags, the ones we intend supporting, as well as the different number of ways that we intend to support.

Otherwise, we're only testing the default layout.